src/conversions/osstr.rs

```diff
@@ -0,0 +1,203 @@
+use crate::types::PyString;
+#[cfg(windows)]
+use crate::PyErr;
+use crate::PyNativeType;
+use crate::{
+    ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
+    ToPyObject,
+};
+use std::borrow::Cow;
+use std::ffi::{OsStr, OsString};
+#[cfg(not(windows))]
+use std::os::raw::c_char;
+
+impl ToPyObject for OsStr {
+    fn to_object(&self, py: Python) -> PyObject {
+        // If the string is UTF-8, take the quick and easy shortcut
+        if let Some(valid_utf8_path) = self.to_str() {
+            return valid_utf8_path.to_object(py);
+        }
+
+        // All targets besides windows support the std::os::unix::ffi::OsStrExt API:
+        // https://doc.rust-lang.org/src/std/sys_common/mod.rs.html#59
+        #[cfg(not(windows))]
+        {
+            let bytes = std::os::unix::ffi::OsStrExt::as_bytes(self);
+
+            let ptr = bytes.as_ptr() as *const c_char;
+            let len = bytes.len() as ffi::Py_ssize_t;
+            unsafe {
+                // DecodeFSDefault automatically chooses an appropriate decoding mechanism to
+                // parse os strings losslessly (i.e. surrogateescape most of the time)
+                let pystring = ffi::PyUnicode_DecodeFSDefaultAndSize(ptr, len);
+                PyObject::from_owned_ptr(py, pystring)
+            }
+        }
+
+        #[cfg(windows)]
+        {
+            let wstr: Vec<u16> = std::os::windows::ffi::OsStrExt::encode_wide(self).collect();
+
+            unsafe {
+                // This will not panic because the data from encode_wide is well-formed Windows
+                // string data
+                py.from_owned_ptr::<PyString>(ffi::PyUnicode_FromWideChar(
+                    wstr.as_ptr(),
+                    wstr.len() as ffi::Py_ssize_t,
+                ))
+                .into()
+            }
+        }
+    }
+}
+
+// There's no FromPyObject implementation for &OsStr because albeit possible on Unix, this would
+// be impossible to implement on Windows. Hence it's omitted entirely
+
+impl FromPyObject<'_> for OsString {
+    fn extract(ob: &PyAny) -> PyResult<Self> {
+        #[cfg(not(windows))]
+        {
+            let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+            // Decode from Python's lossless bytes string representation back into raw bytes
+            let fs_encoded_bytes: &crate::types::PyBytes = unsafe {
+                ob.py()
+                    .from_borrowed_ptr(ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
+            };
```

Comment:
`PyUnicode_EncodeFSDefault` returns a new reference, so we need to own this pointer or we leak memory. I suggest using `Py<PyBytes>`.

```suggestion
            let fs_encoded_bytes: Py<crate::types::PyBytes> = unsafe {
                Py::from_owned_ptr(ob.py(), ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
            };
```

Reply:
I will apply the suggestion. If you have time, I would be interested to hear why you couldn't just replace `from_borrowed_ptr` with `from_owned_ptr`

Reply:
Sure thing. `from_owned_ptr` returns `&PyAny` (or another native type) where the owned pointer has to be stored by PyO3 in a thread-local vector. In comparison, `Py` (and `PyObject`, aka `Py<PyAny>`) directly hold the owned pointer inside them. This is marginally more efficient, and also means that the temporary `bytes` will be cleaned up immediately, instead of when PyO3 has a chance to cleanup its internal vector safely.

This has been a long thorn in PyO3's API imo - there's discussion at #1056 and #1308 where I hope to eventually remove this difference and make everything as efficient as possible!